File: tools/perft.py
"""Perft: exhaustive leaf-node counts for move generation.

Perft walks the full game tree to a fixed depth and counts the leaves. The
number is a fingerprint of the move generator and the rules: if a refactor
changes any legality decision anywhere, the count changes. That makes it the
contract to hold fixed while optimising move generation, where correctness is
otherwise easy to break silently and hard to notice in play.

It is also a clean speed benchmark, because it exercises make/unmake and move
generation with no evaluation, search heuristics or time control involved.

Leaf conventions (kept explicit so the frozen counts are unambiguous):

* A position whose game is already decided — den entry or capture-all, i.e.
  ``GameState.result is not None`` — is a leaf, counted as 1, and is not
  expanded. Continuing past a win would count illegal continuations.
* A position with no legal moves is a leaf, counted as 1. The side to move has
  lost, so there is nothing to expand.
* Otherwise ``perft(0) == 1`` and ``perft(1) == len(legal_moves)``.
"""

from __future__ import annotations

import time

from engine.game_state import GameState


def perft(gs: GameState, depth: int) -> int:
    """Return the number of leaves in the tree of depth *depth* below *gs*."""
    if gs.result is not None:
        return 1
    if depth <= 0:
        return 1
    moves = gs.legal_moves()
    if not moves:
        return 1
    if depth == 1:
        return len(moves)

    total = 0
    for move in moves:
        gs.apply_move(move)
        total += perft(gs, depth - 1)
        gs.undo_move()
    return total


def perft_divide(gs: GameState, depth: int) -> list[tuple[str, int]]:
    """Return per-root-move leaf counts, for locating a perft mismatch.

    When a total diverges, the differing root move narrows the search to one
    subtree instead of the whole game.
    """
    if depth < 1:
        raise ValueError("divide needs depth >= 1")
    out: list[tuple[str, int]] = []
    for move in gs.legal_moves():
        label = f"{move.fc},{move.fr}->{move.tc},{move.tr}"
        gs.apply_move(move)
        out.append((label, perft(gs, depth - 1)))
        gs.undo_move()
    out.sort()
    return out


def timed_perft(gs: GameState, depth: int) -> tuple[int, float]:
    """Return (leaves, seconds) for one perft run."""
    t0 = time.perf_counter()
    nodes = perft(gs, depth)
    return nodes, time.perf_counter() - t0

File: rust/crates/jungle-core/src/perft.rs
//! Exhaustive leaf counts — the move-generation contract.
//!
//! Leaf conventions are copied exactly from `tools/perft.py`, because the frozen
//! counts in `tests/test_perft.py` only mean anything if both implementations
//! agree on what a leaf is:
//!
//! * A decided position (den entry or capture-all) is a leaf, counted as 1, and
//!   is not expanded — continuing past a win would count illegal continuations.
//! * A position with no legal move is a leaf, counted as 1.
//! * Otherwise `perft(0) == 1` and `perft(1) == number of legal moves`.

use crate::movegen::generate;
use crate::position::Position;
use crate::types::Move;

pub fn perft(pos: &mut Position, depth: u32) -> u64 {
    if pos.result().is_some() {
        return 1;
    }
    if depth == 0 {
        return 1;
    }
    let moves = generate(pos);
    if moves.is_empty() {
        return 1;
    }
    if depth == 1 {
        return moves.len() as u64;
    }

    let mut total = 0;
    for &m in moves.as_slice() {
        pos.make(m);
        total += perft(pos, depth - 1);
        pos.unmake();
    }
    total
}

/// Per-root-move counts, for localising a mismatch to one subtree.
pub fn perft_divide(pos: &mut Position, depth: u32) -> Vec<(Move, u64)> {
    assert!(depth >= 1, "divide needs depth >= 1");
    let mut out = Vec::new();
    for &m in generate(pos).as_slice() {
        pos.make(m);
        out.push((m, perft(pos, depth - 1)));
        pos.unmake();
    }
    out.sort_by_key(|&(m, _)| m);
    out
}
